addons/web/controllers/home.py

```diff
@@ -83,9 +83,11 @@ def web_login(self, redirect=None, **kw):
         if request.httprequest.method == 'GET' and redirect and request.session.uid:
             return request.redirect(redirect)
 
-        # so it is correct if overloaded with auth="public"
-        if not request.uid:
-            request.update_env(user=odoo.SUPERUSER_ID)
+        # simulate auth=public, despite having auth=none to be available
+        # to redirect users when no db is selected - cfr ensure_db()
+        if request.env.uid is None:
+            public_user = request.env.ref('base.public_user')
```

Comment:
This will force the wrong public user in multi-website/multi-company environment, leading to 403 errors.
Through the `auth="public"` we actually did set the "correct" website's company's public user, see https://github.com/odoo/odoo/blame/fda28e373c6bacc3a1bccf6c73b83b95d14885f4/addons/website/models/ir_http.py#L126-L137

Step to reproduce on runbot:
- For website 1, set the domain to runbot domain with `-all`
- For website 2, set the domain to runbot domain without `-all`
- Set Chicago company for website 2 and SF company for website 1
- For both website, still in the settings, unselect "Shared Customer Accounts"
- ^ all of this in done through the settings
- Access the runbot URL without `-all`, it will crash, most likely because if forced the wrong public user (the one from the other website's company).

We should probably leabe the `auth=public` in website or go through a util method to get the public user which is overriden in website?

Reply:
Good point! How about we specifically call `_auth_method_public()` then, since it's properly overridden by `website`?

Also we should force this only when there is no logged in user at all, otherwise we do want to see the currently logged user. At least that was the behavior before and that's closer to the behavior of a true `auth=public` route. I've updated the PR with an attempt at both.

BTW I do think an overriding route should never modify an attribute from the original route, as that's an intrusive change for the original module, breaking encapsulation principles. Even here where we have a hybrid configuration:
- the desired behavior is to have `/web/login` behave as a `auth=public` route all the time, so that should be true for `web` even without `website` installed
- changing the route to a real `auth=public` implies that the route should not be visible anymore in "nodb" context, e.g. when there is no db_filter and no user. That's not what we want here.

Reply:
>  How about we specifically call _auth_method_public() then, since it's properly overridden by website?

Seems about right. Working fine regarding the multi-website/multi-company case.

> BTW I do think an overriding route should never modify an attribute from the original route

Agreed.